{code}
package test

abstract class SymbolLoaders {
  val global : Global
  abstract class Loader { def complete(sym : global.Symbol) : Unit }
  class MSILLoader extends Loader {
    private object typeParser extends CLRTypes {
      val global : SymbolLoaders.this.global.type = SymbolLoaders.this.global
    }
    def complete(sym : global.Symbol) = {
      Console.println("completing " + sym)
      typeParser.parseType(sym.asInstanceOf)
    }
  }
  def loader : Loader = new MSILLoader
}
class Global {
  class Symbol
  val loaders = new SymbolLoaders {
    val global : Global.this.type = Global.this
  }
}
abstract class CLRTypes {
  val global : Global
  def parseType(root : global.Symbol) = {}
}


object Go {
  def main(args : Array[String]) : Unit = {
    val global = new Global
    val sym = new global.Symbol
    global.loaders.loader.complete(sym)
  }
}
{code}

On run:

{code}
completing test.Global$$Symbol@6b97fd
Exception in thread "main" java.lang.ClassCastException: test.Global$$Symbol
	at test.SymbolLoaders$$MSILLoader.complete(Go.scala:12)
	at test.Go$$.main(Go.scala:33)
	at test.Go.main(Go.scala)
{code}

Note that without the "asInstanceOf", the code is statically type safe and will run without a run time error.


Another code contribution reaped from the mailing list:

{code}
package test3
object Test extends Application
{
   val a = Array(10, 20, 30)
   (a:AnyRef) match {
   case a : RandomAccessSeq.Mutable[Int] => a
   }
   Console.println("hello")
}
{code}

Changing the priority to high.


For the life of me I can't figure out exactly what bug is being reported here or whether it's still true.  I have opened ticket SI-1754 to report that .asInstanceOf does not do type inference correctly.  If there is some other bug being reported here, I would appreciate someone reopening the ticket and explaining what it is.
